service/resource/deployment/resource.go

```diff
@@ -0,0 +1,294 @@
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty.")
+	}
+
+	newService := &Resource{
+		azureConfig: config.AzureConfig,
+		logger: config.Logger.With(
+			"resource", Name,
+		),
+	}
+
+	return newService, nil
+}
+
+// GetCurrentState gets the current deployments for this cluster from the
+// Azure API.
+func (r *Resource) GetCurrentState(obj interface{}) (interface{}, error) {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	resourceGroupName := key.ClusterID(customObject)
+	deploymentClient, err := r.getDeploymentsClient()
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var deployments []Deployment
+	{
+		for _, deploymentName := range getDeploymentNames() {
+			deploymentExtended, err := deploymentClient.Get(resourceGroupName, deploymentName)
+			if err != nil {
+				if client.ResponseWasNotFound(deploymentExtended.Response) {
+					// Fall through.
+					continue
+				}
+
+				return nil, microerror.Mask(err)
+			}
+
+			deployment := Deployment{
+				Name:            *deploymentExtended.Name,
+				Parameters:      *deploymentExtended.Properties.Parameters,
+				ResourceGroup:   resourceGroupName,
+				TemplateURI:     *deploymentExtended.Properties.TemplateLink.URI,
+				TemplateVersion: *deploymentExtended.Properties.TemplateLink.ContentVersion,
+			}
+			deployments = append(deployments, deployment)
+		}
+	}
+
+	return deployments, nil
+}
+
+// GetDesiredState returns the desired deployments for this cluster.
+func (r *Resource) GetDesiredState(obj interface{}) (interface{}, error) {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterSetup, err := newClusterSetupDeployment(customObject)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	deployments := []Deployment{
+		clusterSetup,
+	}
+	return deployments, nil
+}
+
+// GetCreateState returns the deployments that should be created for this
+// cluster.
+func (r *Resource) GetCreateState(obj, currentState, desiredState interface{}) (interface{}, error) {
+	currentDeployments, err := toDeployments(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredDeployments, err := toDeployments(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var deploymentsToCreate []Deployment
+
+	for _, desiredDeployment := range desiredDeployments {
+		if !existsDeploymentByName(currentDeployments, desiredDeployment.Name) {
+			deploymentsToCreate = append(deploymentsToCreate, desiredDeployment)
+		}
+	}
+
+	return deploymentsToCreate, nil
+}
+
+// GetDeleteState returns an empty deployments collection. Deployments and the
+// resources they manage are deleted when the Resource Group is deleted.
+func (r *Resource) GetDeleteState(obj, currentState, desiredState interface{}) (interface{}, error) {
```

Comment:
See the comment. We delete the resource group to ensure all resources are deleted.